[RFC] Improve errors for fields overriding destructor attributes

MoonlightSentinel opened the pull request:
Compiler generated destructors calling both field and user-defined
destructors can have less-restrictive attributes than the user-defined
one. This caused confusing error messages stating that the destructor
is not `pure`albeit the user marked their implementation as `pure`.

This commit adds an additional error message explaining the situation
and also lists the fields causing the attribute missmatch s.t. the user
may fix their respective destructors.

---

TL;DR: The compiler tells you why your code doesn't compile even though your non-rewritten code should - especially important for `-preview=fielddtors`.

MoonlightSentinel committed:
Improve errors for fields overriding destructor attributes

Compiler generated destructors calling both field and user-defined
destructors can have less-restrictive attributes than the user-defined
one. This caused confusing error messages stating that the destructor
is not `pure`albeit the user marked their implementation as `pure`.

This commit adds an additional error message explaining the situation
and also lists the fields causing the attribute missmatch s.t. the user
may fix their respective destructors.

kinke commented:
Completely unrelated, but with the introduction of `frontend.h`, every new `extern(D)` function addition requiring a change and me being a code-owner of *.h, I will now probably be requested for review for almost every PR. Can we get rid of these `// ignoring extern () block because of linkage` comments (which don't even mention the name of the function and are so IMO hardly useful anyway)?

MoonlightSentinel commented:
Yeah, that's annoying to update as well.

wilzbach commented on src/dmd/expression.d:
```diff
@@ -1107,11 +1107,81 @@ extern (C++) abstract class Expression : ASTNode
+     *   sc = scope
+     *   f  = potential `DtorDeclaration`
+     *   check = current check (e.g. whether it's pure)
+     *   what = the kind of check (e.g. `"pure"`)
```
```suggestion
     *   checkName = the kind of check (e.g. `"pure"`)
```

wilzbach commented on test/fail_compilation/dtor_attributes.d:
```diff
@@ -0,0 +1,145 @@
+
+struct StrictNested
+{
+    HasDtor[4] arrayMember;
```
Could you also add `HasDtorClass[4] classMember;` (for completeness)

wilzbach commented on test/fail_compilation/dtorfields_attributes.d:
```diff
@@ -0,0 +1,42 @@
+fail_compilation\dtorfields_attributes.d(117):         - member
+---
+
+TODO: nothrow is not checked, but that seems to be an unrelated bug
```
Better open a bugzilla issue and link it here (otherwise it gets forgotten)

wilzbach commented on src/dmd/expression.d:
```diff
@@ -1107,11 +1107,81 @@ extern (C++) abstract class Expression : ASTNode
+            // Sanity check
+            assert(!check(cast(DtorDeclaration) ad.fieldDtor));
+
+            errorSupplemental("it's %s because of the following field's destructors:", what);
```
```suggestion
            errorSupplemental("it's `%.*s` because of the following field's destructors:", checkName.length, checkName.ptr);
```

wilzbach commented on src/dmd/expression.d:
```diff
@@ -1107,11 +1107,81 @@ extern (C++) abstract class Expression : ASTNode
+     *   what = the kind of check (e.g. `"pure"`)
+     */
+    extern (D) void checkOverridenDtor(Scope* sc, FuncDeclaration f,
+                scope bool function(DtorDeclaration) check, const char* what
```
```suggestion
                scope bool function(DtorDeclaration) check, string checkName
```

(new code should prefer to use D strings)

wilzbach reviewed (approved):
This will be a great help to the users trying to decipher the error messages!

I wonder whether it would make sense (for an extension/future work) to generalize this as deciphering a `@system` chain can often be hard if templates are involved.

MoonlightSentinel commented:
> I wonder whether it would make sense (for an extension/future work) to generalize this as deciphering a `@system` chain can often be hard if templates are involved.

You mean something like this:

```
import core.stdc.stdio : puts;

void foo()()
{
     puts("Hello, World!");
}

void main() @safe
{
    foo!();
}
```

```
onlineapp.d(10): Error: @safe function D main cannot call @system function onlineapp.foo!().foo
onlineapp.d(3):        onlineapp.foo!().foo is declared here
onlineapp.d(3):        onlineapp.foo!().foo is inferred @system because of the following operations:
onlineapp.d(5):        -  puts("Hello, World!");
```

MoonlightSentinel committed:
Add destructor tracking

Backtrace through multiple destructors to find the culprit why
a generated destructor is inferred e.g. @system.

MoonlightSentinel commented:
Added tracking for destructors s.t. you see the desctructor of a member of a member .... which causes the less-restrictive attribute inference.

nordlow commented:
This is awesome diagnostics. How hard would it be to do the same to trace transitive mismatches for qualifiers `@safe`, `pure`, `nothrow`, `@nogc`?

MoonlightSentinel commented:
This diagnostic was rather easy because we know exactly what t look for. A general check should be possible using a visitor but are propbably a lot harder to get right.
